### PR TITLE
Refactor specific docker image tests to push common behaviour to an interface/abstract class and inherit from it

### LINKED ISSUE #22: Refactor specific docker image tests to push common behaviour to an interface/abstract class and inherit from it
Waiting on #3 to be closed.

Currently the following tests have the same layout:
- SQLServerDockerImageTest
- PostgreSQLDockerImageTest
- MongoDBDockerImageTest
- MySQLDockerImageTest
- SplunkDockerImageTest
- ElasticSearchDockerImageTest

All of the above mentioned test classes have a single test `testXXXDockerImage()` which ends up calling a `connectToXXXDatabase()` to ensure that a successful connection has been made.

We can rename these methods to something more generic like:
`testXXXDockerImage()` -> `testDockerImage()`
`connectToXXXDatabase()` -> `testServiceProvidedByDockerContainer()`

And we can have an interface called DockerImageTestInterface where:
- We have a default implementation of `testDockerImage()` calling `testServiceProvidedByDockerContainer()`
- We have an abstract declaration of `testServiceProvidedByDockerContainer()` which is to be defined by the implementing class.

And all the tests can implement `DockerImageTestInterface` and it's own `testServiceProvidedByDockerContainer()`.

The above proposed change, will reduce code redundancy and make it easier to maintain these tests.

### PR DESCRIPTION
**Pull request by the-c0d3br34k3r**
Closes #22.
Previously all the container specific tests have duplicate deployment code.
Now, the common code has been pushed to an abstract class from which all the
container specific tests inherit.

### CODE REVIEW

File: src/test/java/SplunkDockerImageTest.java [hunk 1]
```diff
@@ -28,33 +33,7 @@ private void connectToSplunk(DockerImage dockerImage) {
 
         System.setProperty("https.protocols", "TLSv1,TLSv1.1,TLSv1.2");
 
-        // Create a Service instance and log in with the argument map
-        Service service = Service.connect(loginArgs);
-
-        System.out.println(service.getApplications());
```
Comment (abprash) [inline src/test/java/SplunkDockerImageTest.java:34]: why was this removed?

File: src/test/java/SQLServerDockerImageTest.java [hunk 2]
```diff
@@ -31,49 +43,22 @@ private void connectToMSSQLDatabase(DockerImage dockerImage) throws SQLException
...
-            // Create and run container
-            dockerUtility.createAndRunContainer(dockerImage, containerName, true);
-
-            // Wait 20s for the SQL server to start up
-            Thread.sleep(20_000);
-
-            // Test if the connection to the database is successful
-            Assertions.assertDoesNotThrow(() -> connectToMSSQLDatabase(dockerImage));
-        }
-        finally {
```
Comment (abprash) [inline src/test/java/SQLServerDockerImageTest.java:74]: we should have cleanup logic?

Comment (abprash) [reply src/test/java/SQLServerDockerImageTest.java:74]: nvm saw its in the abstract class.

File: src/main/java/DockerImage.java [hunk 3]
```diff
@@ -106,6 +106,22 @@ public String getRepository() {
         return this.environmentVariables;
     }
 
+    /**
+     * Returns the value of the specified environment variable if it is configured for this Docker image,
+     * else returns a null.
```
Comment (abprash) [inline src/main/java/DockerImage.java:111]: can we return an empty string ? Everywhere we used this method, we don't really check if its null.

File: src/test/java/ADockerImageTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public abstract class ADockerImageTest {
+
+    /**
+     * Returns a list of docker images to deploy as containers and test.
```
Comment (abprash) [inline src/test/java/ADockerImageTest.java:7]: duplicated comments ?

File: src/main/java/DockerImage.java (hunk 3 above)
Comment (the-c0d3br34k3r) [reply src/main/java/DockerImage.java:111]: The reason I chose to return a NULL is because, there can be env variables with empty values, which will be mapped to an empty string.

File: src/test/java/ADockerImageTest.java (hunk 4 above)
Comment (the-c0d3br34k3r) [reply src/test/java/ADockerImageTest.java:7]: As in? What is duplicated here?

File: src/test/java/SplunkDockerImageTest.java (hunk 1 above)
Comment (the-c0d3br34k3r) [reply src/test/java/SplunkDockerImageTest.java:34]: If you're asking about the sysout statement, just cleaned it as we are touching this file.
And since we aren't printing the applications, we don't need the service variable anymore, just a call to Service::connect() would suffice - https://github.com/padaiyal/jAdya/pull/24/files/2aa53544dfa59d646ce88178e23ce265a682f3d6#diff-de09bfdbf962b08e15f78e58d544e74f9dbcc9b40a4232f6426a3005450d6907R37

File: src/test/java/ADockerImageTest.java (hunk 4 above)
Comment (abprash) [reply src/test/java/ADockerImageTest.java:7]: Line 7&8.

File: src/main/java/DockerImage.java (hunk 3 above)
Comment (abprash) [reply src/main/java/DockerImage.java:111]: Got it. Makes sense. Do you think we need to have NPE checks when we use env variables?

File: src/test/java/SplunkDockerImageTest.java (hunk 1 above)
Comment (abprash) [reply src/test/java/SplunkDockerImageTest.java:34]: Got it. So connect() would suffice in testing out the container is fully up and running?
I wanted to make it uniform, since we do some operations in other containers to test. Let me know what you think.

Comment (the-c0d3br34k3r) [reply src/test/java/SplunkDockerImageTest.java:34]: Post out discussion, we decided to keep it as is i.e. connect() would suffice to test if the container is up and running.

File: src/main/java/DockerImage.java (hunk 3 above)
Comment (the-c0d3br34k3r) [reply src/main/java/DockerImage.java:111]: Added null checks to retrieved env variables.

File: src/test/java/ADockerImageTest.java (hunk 4 above)
Comment (the-c0d3br34k3r) [reply src/test/java/ADockerImageTest.java:7]: Post discussion, keeping as is.

**Review by abprash — APPROVED** [review]
LGTM
